File: Hard/2197. Replace Non-Coprime Numbers in Array/solution.py
class Solution:
    def replaceNonCoprimes(self, nums: List[int]) -> List[int]:
        stack = []
        current = nums[0]
        for num in nums[1:]:
            if self.GCD(current, num) > 1:
                current = self.LCM(current, num)
                while stack and self.GCD(current, stack[-1]) > 1:
                    current = self.LCM(current, stack.pop())
            else:
                stack.append(current)
                current = num

        stack.append(current)
        return stack
    
    def GCD(self, a, b):
        if a == 1 or b == 1: return 1
        while b > 0:
            tmp = b
            b = a % b
            a = tmp
        return a
    
    def LCM(self, a, b):
        return int(a * b / self.GCD(a, b))

File: Hard/2197. Replace Non-Coprime Numbers in Array/solution.cs
public class Solution {
    public IList<int> ReplaceNonCoprimes(int[] nums) {
        Stack<int> stack = new Stack<int>();
        int current = nums[0];
        for(int i = 1; i < nums.Length; i++){
            int gcd = GCD(current, nums[i]);
            if(gcd > 1){
                current = current / gcd * nums[i];

                while(stack.Count > 0 && GCD(current, stack.Peek()) > 1){
                    current = LCM(current, stack.Pop());
                }
            } else{
                stack.Push(current);
                current = nums[i];
            }
        }
        stack.Push(current);
        
        IList<int> res = stack.Reverse().ToList();
        return res;
    }

    public int GCD(int a, int b){
        if(a == 1 || b == 1) return 1;
        while(b > 0){
            int tmp = b;
            b = a % b;
            a = tmp;
        }
        return a;
    } 

    public int LCM(int a, int b){
        return a / GCD(a, b) * b;
    }
}